src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -126,30 +119,99 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
    * @param thisNode the input data Node object
    * @return a useful node used to construct the Guava Graph
    */
-  private GraphNode protoNodeToGraphNode(Node thisNode) {
-    return GraphNode.create(thisNode.getName(), thisNode.getTokenList(), thisNode.getMetadata());
+  public GraphNode protoNodeToGraphNode(Node thisNode) {
+    List<String> newTokenList = new ArrayList<>();
+    newTokenList.addAll(thisNode.getTokenList());
+    Struct newMetadata = Struct.newBuilder().mergeFrom(thisNode.getMetadata()).build();
+    return GraphNode.create(thisNode.getName(), newTokenList, newMetadata);
+  }
+
+  /*
+   * Changes the graph according to the given mutation object
+   * @param mut the mutation to affect
+   * @param graph the Guava graph to mutate
+   * @param graphNodesMap a reference of existing nodes, also to be mutated
+   * @return true if the mutation was successful, false otherwise
+   */
+  public boolean mutateGraph(
+      Mutation mut, MutableGraph<GraphNode> graph, Map<String, GraphNode> graphNodesMap) {
+    // Nodes affected by the mutation
+    // second node only applicable for adding an edge and removing an edge
+    String startName = mut.getStartNode();
+    String endName = mut.getEndNode();
+
+    // Getting the corresponding graph nodes from the graph map
+    GraphNode startNode = graphNodesMap.get(startName);
+    GraphNode endNode = graphNodesMap.get(endName);
+
+    switch (mut.getType()) {
+      case ADD_NODE:
+        if (graphNodesMap.containsKey(startName)) {
+          // adding a duplicate node
+          return false;
+        }
+        // Create a new node with the given name and add it to the graph and the map
+        GraphNode newGraphNode =
+            GraphNode.create(startName, new ArrayList<>(), Struct.newBuilder().build());
+        graph.addNode(newGraphNode);
+        graphNodesMap.put(startName, newGraphNode);
+        break;
+      case ADD_EDGE:
+        if (startNode != null && endNode != null) { // Check nodes exist before adding an edge
```

Comment:
In here you have a lot of patterns like
```
case A:
   if (condition) {
         do something
         // expect to return true at the end of the function.
   } else {
         return false;
   }
```
which I think would be more easy to follow if you did:
```
case A:
    if (! condition) {
         return false;
    }
    do something
```